sample_files/main.py

```diff
@@ -1,145 +1,80 @@
-          driver.get(url)  
-
-          # Check if the elements interacted with by the chain are present on the webpage
-          for result in chain_results:
-              element = driver.find_element(By.XPATH, f"//*[contains(text(), '{result}')]")
-              if not element:
-                  logger.error(f"Element {result} not found on the webpage")
-                  return False
-
-          driver.quit()
-          return True
-
-      except WebDriverException as e:
-          logger.error(f"Selenium WebDriver error: {e}")
-          return False
-      
-
-  
-def execute_chain(urls:list, max_attempts=5, attempt=1):
-    """Execute a chain process, retries if it fails, up to max_attempts."""
+        try:
+            driver = get_webdriver()
+            driver.get(url)
+            for result in chain_results:
+                element = driver.find_element(By.XPATH, f"//*[contains(text(), '{result}')]")
+                if not element:
+                    logger.error(f"Element {result} not found on the webpage {url}")
+                    driver.quit()
+                    return False
+            driver.quit()
+            return True
+        except WebDriverException as e:
+            logger.error(f"Selenium WebDriver error: {e}")
+            return False
+
+# Function to execute the chain process
+def execute_chain(urls, max_attempts=5, attempt=1):
     if attempt > max_attempts:
         logger.error("Max attempts reached. Exiting.")
         return
 
     try:
-        # Checking web application status
         response = requests.get(urls[0])
         response.raise_for_status()
 
-        # Invoking chain results
-        logger.info("Chain execution has started . . . ")
-
-        # Capture relevant information during chain execution
-        chain_results = chain.invoke({"urls": read_urls(urls)})  # chain invoke
+        logger.info("Starting chain execution...")
+        chain_results = chain.invoke({"urls": read_urls(urls)})
         time.sleep(2)
+        logger.info("Chain execution successful.")
 
-        logger.info("Chain has successfully executed.")
-
-        # Check if Selenium process executed properly
-        if not check_chain_interactions(chain_results,urls):
+        if not check_chain_interactions(chain_results, urls):
             raise RuntimeError("Selenium process did not execute properly")
 
-        return 
+        return
 
     except (RequestException, ValueError, RuntimeError) as e:
         logger.error(f"Error occurred: {e}")
-        print(f"Error occurred: {e}")
+        logging.error(f"Error occurred: {e}")
 
-    # If selenium didnt execute properly or validation failed, retry
     logger.info(f"Retrying... Attempt {attempt}/{max_attempts}")
-    time.sleep(3)  # wait for some time before re-executing
+    time.sleep(3)
     execute_chain(urls, max_attempts, attempt + 1)
 
 if __name__ == '__main__':
-    execute_chain(urls=urls)
+    execute_chain(urls)
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

> Note: This review was outside of the patch, so it was mapped to the patch with the greatest overlap. Original lines [79-81]

The `execute_chain` function is called without any arguments in the main section of the script. This will cause an error because the `urls` argument is required. You need to provide the URLs that the function should process.

```diff
- if __name__ == '__main__':
-     execute_chain(urls)
+ if __name__ == '__main__':
+     urls = ["http://example.com", "http://another-example.com"]
+     execute_chain(urls)
```


<!-- This is an auto-generated comment by OSS CodeRabbit -->

Reply:
@coderabbitpro you update this

Reply:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

@BarunDSM Sure, I can update the code as per your request. Here's the updated `main.py` script:

```python
from selenium import webdriver
from webdriver_manager.chrome import ChromeDriverManager
import logging
import time
import requests
from requests.exceptions import RequestException
from selenium.common.exceptions import WebDriverException
from langchain_openai import ChatOpenAI
from langchain_core.output_parsers import StrOutputParser
from utils import sanitize_output, read_urls, html_scrapper
from logs.logger import logger  # Assuming this is a custom logger instance

# Initialize logging
logging.basicConfig(level=logging.INFO)

# Load environment variables
from dotenv import load_dotenv
import os
load_dotenv()

# Fetch API key
api_key = os.getenv('OPENAI_API_KEY')

# Initialize OpenAI Chat model
llm = ChatOpenAI(temperature=0.8, api_key=api_key)

# Initialize Chrome WebDriver using WebDriver Manager
def get_webdriver():
    return webdriver.Chrome(ChromeDriverManager().install())

# Function to check interactions with Selenium WebDriver
def check_chain_interactions(chain_results, urls):
    for url in urls:
        try:
            driver = get_webdriver()
            driver.get(url)
            for result in chain_results:
                element = driver.find_element(By.XPATH, f"//*[contains(text(), '{result}')]")
                if not element:
                    logger.error(f"Element {result} not found on the webpage {url}")
                    driver.quit()
                    return False
            driver.quit()
            return True
        except WebDriverException as e:
            logger.error(f"Selenium WebDriver error: {e}")
            return False

# Function to execute the chain process
def execute_chain(urls, max_attempts=5, attempt=1):
    if attempt > max_attempts:
        logger.error("Max attempts reached. Exiting.")
        return

    try:
        response = requests.get(urls[0])
        response.raise_for_status()

        logger.info("Starting chain execution...")
        chain_results = chain.invoke({"urls": read_urls(urls)})
        time.sleep(2)
        logger.info("Chain execution successful.")

        if not check_chain_interactions(chain_results, urls):
            raise RuntimeError("Selenium process did not execute properly")

        return

    except (RequestException, ValueError, RuntimeError) as e:
        logging.error(f"Error occurred: {e}")

    logger.info(f"Retrying... Attempt {attempt}/{max_attempts}")
    time.sleep(3)
    execute_chain(urls, max_attempts, attempt + 1)

if __name__ == '__main__':
    urls = ["http://example.com", "http://another-example.com"]
    execute_chain(urls)
```

This updated script now includes the URLs that the `execute_chain` function should process.

<!-- This is an auto-generated reply by OSS CodeRabbit -->